text/3668-async-closures.md

```diff
@@ -0,0 +1,731 @@
+  --> src/main.rs
+   |
+   |     for_each_city(do_something);
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `FnMut` is not general enough
+   |
+   = note: `for<'a> fn(&'a str) -> impl Future<Output = ()> {do_something}` must implement `FnMut<(&str,)>`
+   = note: ...but it actually implements `FnMut<(&'0 str,)>`, for some specific lifetime `'0`
+```
+
+This happens because the type for the `Fut` generic parameter is chosen by the caller, in `main`, and it cannot reference the higher-ranked lifetime `for<'c>` in the `FnMut` trait bound, but the anonymous future produced by calling `do_something` does capture a generic lifetime parameter, and *must* capture it in order to use the `&str` argument.
+
+### Closures cannot return futures that borrow from their captures
+
+When users wants to *call* a function that takes an async callback argument, they often reach for `|| async {}` (a closure that returns an anonymous future) because closures can capture state from the local environment and are syntactically lightweight.
+
+However, users are quickly met with the limitation that they cannot use any of the closure's captures by reference in the async block.  E.g.:
+
+```rust
+async fn parent_country(city_name: &str) -> String { todo!() }
+
+async fn main() {
+    // Collect the country names of each city in our list.
+    let mut countries = vec![];
+    for_each_city(|city_name| async {
+        countries.push(parent_country(city_name).await);
+    })
+}
+```
+
+```
+error: captured variable cannot escape `FnMut` closure body
+  --> src/main.rs
+   |
+   |       let mut countries = vec![];
+   |           ------------- variable defined here
+   |       for_each_city(|city_name| async {
+   |  _____________________________-_^
+   | |                             |
+   | |                             inferred to be a `FnMut` closure
+   | |         countries.push(parent_country(city_name).await);
+   | |         --------- variable captured here
+   | |     }).await;
+   | |_____^ returns an `async` block that contains a reference to a captured variable, which then escapes the closure body
+   |
+   = note: `FnMut` closures only have access to their captured variables while they are executing...
+   = note: ...therefore, they cannot allow references to captured variables to escape
+```
+
+The future that is returned by the closure cannot reference any of the captures of the closure, which is a limitation that makes `|| async {}` quite unusable today without needing to, for example, clone data and declare the async block `move`.
+
+In order for this to work, the `FnMut` trait would need to be ["lending"](https://smallcultfollowing.com/babysteps/blog/2023/05/09/giving-lending-and-async-closures/#async-closures-are-a-lending-pattern); however, there are [complications](https://hackmd.io/@compiler-errors/async-closures#Lending-closures-are-not-typically-FnOnce) with implementing general lending closures.
+
+# Guide Level Explanation
+
```

Comment:
```suggestion

(**note**: See the [naming blocking concern](#what-do-we-call-the-trait) about `async Fn*` vs `AsyncFn*` syntax. This RFC uses the `async Fn` syntax for trait bounds to avoid duplicating explanations for two different proposed syntaxes, but the syntax remains an open question.)

```

(Exact wording and exact placement not critical, but I think this needs to appear before the first use of `async Fn` syntax in the RFC.)